src/data/nat/modeq.lean

```diff
@@ -105,32 +105,57 @@ by rw [modeq_iff_dvd] at *; exact dvd.trans (dvd_mul_left (n : ℤ) (m : ℤ)) h
 theorem modeq_of_modeq_mul_right (m : ℕ) : a ≡ b [MOD n * m] → a ≡ b [MOD n] :=
 mul_comm m n ▸ modeq_of_modeq_mul_left _
 
+theorem modeq_one : a ≡ b [MOD 1] := modeq_of_dvd $ one_dvd _
+
 local attribute [semireducible] int.nonneg
 
-/-- The natural number less than `n*m` congruent to `a` mod `n` and `b` mod `m` -/
-def chinese_remainder (co : coprime n m) (a b : ℕ) : {k // k ≡ a [MOD n] ∧ k ≡ b [MOD m]} :=
-⟨let (c, d) := xgcd n m in int.to_nat ((b * c * n + a * d * m) % (n * m)), begin
-  rw xgcd_val, dsimp [chinese_remainder._match_1],
-  rw [modeq_iff_dvd, modeq_iff_dvd],
-  rw [int.to_nat_of_nonneg], swap,
-  { by_cases h₁ : n = 0, {simp [coprime, h₁] at co, substs m n, simp},
-    by_cases h₂ : m = 0, {simp [coprime, h₂] at co, substs m n, simp},
-    exact int.mod_nonneg _
-      (mul_ne_zero (int.coe_nat_ne_zero.2 h₁) (int.coe_nat_ne_zero.2 h₂)) },
-  have := gcd_eq_gcd_ab n m, simp [co.gcd_eq_one, mul_comm] at this,
+/-- The natural number less than `lcm n m` congruent to `a` mod `n` and `b` mod `m` -/
+def chinese_remainder' (h : a ≡ b [MOD gcd n m]) : {k // k ≡ a [MOD n] ∧ k ≡ b [MOD m]} :=
+if hn : n = 0 then ⟨a, begin simp [hn] at *, split, refl, exact h end⟩ else
+if hm : m = 0 then ⟨b, begin simp [hm] at *, split, exact h.symm, refl end⟩ else
+⟨let (c, d) := xgcd n m in int.to_nat (((b * c * n + a * d * m) / gcd n m) % lcm n m), begin
+  rw xgcd_val, dsimp [chinese_remainder'._match_1],
+  rw [modeq_iff_dvd, modeq_iff_dvd, int.to_nat_of_nonneg],
+  swap, { exact int.mod_nonneg _ (int.coe_nat_ne_zero.2 (lcm_ne_zero hn hm)), },
+  have := gcd_eq_gcd_ab n m, simp [mul_comm] at this,
   rw [int.mod_def, ← sub_add, ← sub_add]; split,
-  { refine dvd_add _ (dvd_trans (dvd_mul_right _ _) (dvd_mul_right _ _)),
-    rw [add_comm, ← sub_sub], refine _root_.dvd_sub _ (dvd_mul_left _ _),
-    have := congr_arg ((*) ↑a) this,
-    exact ⟨_, by rwa [mul_add, ← mul_assoc, ← mul_assoc, mul_one, mul_comm,
-        ← sub_eq_iff_eq_add] at this⟩ },
-  { refine dvd_add _ (dvd_trans (dvd_mul_left _ _) (dvd_mul_right _ _)),
-    rw [← sub_sub], refine _root_.dvd_sub _ (dvd_mul_left _ _),
-    have := congr_arg ((*) ↑b) this,
-    exact ⟨_, by rwa [mul_add, ← mul_assoc, ← mul_assoc, mul_one, mul_comm _ ↑m,
-        ← sub_eq_iff_eq_add'] at this⟩ }
+  { refine dvd_add _ _,
```

Comment:
There's a fair amount of repetition in this proof. I haven't opened it in Lean yet, but are there some `have`s or lemmas that can be extracted?

Reply:
I've golfed/reorganised a bit, its still a bit similar in the two cases as there is a symmetry to the proof, but hopefully less outright repetition of subgoals. Marking this as resolved but if you can think of other specifics go ahead!